go/flagutil/enum.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2022 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreedto in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagutil
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+)
+
+// StringEnum provides a string-like flag value that raises an error if given a
+// value not in the set of allowed choices.
+//
+// This parse-time validation can be case-sensitive or not, depending on which
+// constructor (NewStringEnum vs NewCaseInsensitiveStringEnum) was used.
+type StringEnum struct {
+	name string
+	val  string
+
+	caseInsensitive bool
+	choices         map[string]struct{}
+	choiceNames     []string
+	choiceMapper    func(string) string
+}
+
+// ErrInvalidChoice is returned when parsing a value that is not a valid choice
+// for the StringEnum flag.
+var ErrInvalidChoice = errors.New("invalid choice for enum")
+
+// NewStringEnum returns a new string enum flag with the given name, default,
+// and choices.
+//
+// Parse-time validation is case-sensitive.
+func NewStringEnum(name string, initialValue string, choices []string) *StringEnum {
+	return newStringEnum(name, initialValue, choices, false)
+}
+
+// NewCaseInsensitiveStringEnum returns a new string enum flag with the given
+// name, default, and choices.
+//
+// Parse-time validation is case-insensitive.
+func NewCaseInsensitiveStringEnum(name string, initialValue string, choices []string) *StringEnum {
+	return newStringEnum(name, initialValue, choices, true)
+}
```

Comment:
TIL; does vitess use it anywhere outside of the mysql context (i.e. not collations)? I think I'd have a hard time puzzling out what "CI" meant in a function name because it definitely isn't related to test infrastructure 😂